### PR TITLE
Replace HPU with external implementation

### PR DESCRIPTION
**Pull request by Borda**
## What does this PR do?

The HPU Strategy moves to an external repo where it will be maintained: https://github.com/Lightning-AI/lightning-Habana/pull/1. A corresponding package was published to PyPi from which users can install the integration. 

- [x] Replaces the implementation with the one in the `lightning-habana` package. 
- [x] Registeres the `accelerate="hpu"` as a strategy if the package is installed
- [x] Produces an error message if you request the strategy but don't have the package installed

**Before:**

```py
trainer = Trainer(accelerator="hpu")

# OR

from lightning.pytorch.accelartors import HPUAccelerator
trainer = Trainer(strategy=HPUAccelerator(...))
```

**Now:**

```bash
pip install lightning-habana
```

```py
trainer = Trainer(accelerator="hpu")  # only works if  `lightning-habana` is installed.

# OR

from lightning_habana import AcceleratorHPU
trainer = Trainer(strategy=AcceleratorHPU(...))
```

<details>
  <summary><b>Before submitting</b></summary>

- Was this **discussed/agreed** via a GitHub issue? (not for typos and docs)
- [x] Did you read the [contributor guideline](https://github.com/Lightning-AI/lightning/blob/master/.github/CONTRIBUTING.md), **Pull Request** section?
- [x] Did you make sure your **PR does only one thing**, instead of bundling different changes together?
- Did you make sure to **update the documentation** with your changes? (if necessary)
- Did you write any **new necessary tests**? (not for typos and docs)
- [x] Did you verify new and **existing tests pass** locally with your changes?
- Did you list all the **breaking changes** introduced by this pull request?
- Did you **update the CHANGELOG**? (not for typos, docs, test updates, or minor internal changes/refactors)

</details>

## PR review

Anyone in the community is welcome to review the PR.
Before you start reviewing, make sure you have read the [review guidelines](https://github.com/Lightning-AI/lightning/wiki/Review-guidelines). In short, see the following bullet-list:

<details>
  <summary>Reviewer checklist</summary>

- [x] Is this pull request ready for review? (if not, please submit in draft mode)
- [x] Check that all items from **Before submitting** are resolved
- [x] Make sure the title is self-explanatory and the description concisely explains the PR
- [x] Add labels and milestones (and optionally projects) to the PR so it can be classified

</details>

cc @carmocca @akihironitta @borda @justusschock @awaelchli

### CODE REVIEW

File: docs/source-pytorch/accelerators/hpu_basic.rst [hunk 1]
```diff
@@ -1,109 +0,0 @@
-:orphan:
-
-.. _hpu_basics:
```
Comment (Borda) [inline docs/source-pytorch/accelerators/hpu_basic.rst:3]: this was moved to the one-pager

File: .github/workflows/ci-dockers-pytorch.yml [hunk 2]
```diff
@@ -179,43 +179,6 @@ jobs:
         env:
           SLACK_WEBHOOK_URL: ${{ secrets.SLACK_WEBHOOK_URL }}
 
-  build-hpu:
```
Comment (Borda) [inline .github/workflows/ci-dockers-pytorch.yml:182]: building these dockers was moved to the separate repo

File: examples/pytorch/hpu/mnist_sample.py [hunk 3]
```diff
@@ -63,7 +63,7 @@ def configure_optimizers(self):
             "accelerator": "hpu",
             "devices": 1,
             "max_epochs": 1,
-            "plugins": lazy_instance(HPUPrecisionPlugin, precision="16-mixed"),
+            "plugins": lazy_instance(PrecisionHPU, precision="16-mixed"),
```
Comment (awaelchli) [inline examples/pytorch/hpu/mnist_sample.py:66]: Why are all the names backward? It should be HPUPrecision, HPUAccelerator, and so on.

File: src/lightning/pytorch/trainer/connectors/accelerator_connector.py [hunk 4]
```diff
@@ -414,10 +413,20 @@ def _choose_strategy(self) -> Union[Strategy, str]:
         if self._accelerator_flag == "ipu":
             return IPUStrategy.strategy_name
         if self._accelerator_flag == "hpu":
+            if not _LIGHTNING_HABANA_AVAILABLE:
+                raise NotImplementedError(
```
Comment (awaelchli) [inline src/lightning/pytorch/trainer/connectors/accelerator_connector.py:417]:
```suggestion
                raise ImportError(
```

File: src/lightning/pytorch/trainer/setup.py [hunk 5]
```diff
@@ -191,11 +190,19 @@ def _log_device_info(trainer: "pl.Trainer") -> None:
...
 
-    if _HPU_AVAILABLE and not isinstance(trainer.accelerator, HPUAccelerator):
-        rank_zero_warn(
-            "HPU available but not used. Set `accelerator` and `devices` using"
-            f" `Trainer(accelerator='hpu', devices={HPUAccelerator.auto_device_count()})`."
-        )
+    if _HPU_AVAILABLE:
+        if not _LIGHTNING_HABANA_AVAILABLE:
+            raise ModuleNotFoundError(
+                "You are running on HPU machine but you have not installed `lightning-habana` extension."
```
Comment (awaelchli) [inline src/lightning/pytorch/trainer/setup.py:196]:
```suggestion
                f"You are running on HPU machine but you have not installed `lightning-habana` extension. {str(_LIGHTNING_HABANA_AVAILABLE)}"
```

File: src/lightning/pytorch/utilities/testing/_runif.py [hunk 6]
```diff
@@ -23,14 +23,13 @@
...
 from lightning.pytorch.accelerators.cpu import _PSUTIL_AVAILABLE
-from lightning.pytorch.accelerators.hpu import _HPU_AVAILABLE
 from lightning.pytorch.accelerators.ipu import _IPU_AVAILABLE
 from lightning.pytorch.accelerators.mps import MPSAccelerator
 from lightning.pytorch.accelerators.tpu import TPUAccelerator
 from lightning.pytorch.callbacks.progress.rich_progress import _RICH_AVAILABLE
 from lightning.pytorch.core.module import _ONNX_AVAILABLE
 from lightning.pytorch.strategies.deepspeed import _DEEPSPEED_AVAILABLE
-from lightning.pytorch.utilities.imports import _OMEGACONF_AVAILABLE
+from lightning.pytorch.utilities.imports import _HPU_AVAILABLE, _OMEGACONF_AVAILABLE
```
Comment (awaelchli) [inline src/lightning/pytorch/utilities/testing/_runif.py:32]: remove `RunIf(hpu=...)` too?

File: examples/pytorch/hpu/mnist_sample.py (hunk 3 above)
Comment (Borda) [reply examples/pytorch/hpu/mnist_sample.py:66]: hmm I can rever it but found it strange that had `HPUSingleStrategy`
so ho I read it a posix - Accelerator for HPU

Comment (justusschock) [reply examples/pytorch/hpu/mnist_sample.py:66]: I agree with @awaelchli . let's be consistent with the naming we have everywhere else which is a device-prefix

**Review by justusschock — APPROVED** [review]

File: examples/pytorch/hpu/mnist_sample.py (hunk 3 above)
Comment (Borda) [reply examples/pytorch/hpu/mnist_sample.py:66]: ok, will revert it but still would instead do this way :D

**Review by awaelchli — APPROVED** [review]
LGTM

**Review by carmocca — COMMENTED** [review]
CI is red fyi. Pushed some missed removals too.

File: tests/tests_pytorch/trainer/connectors/test_accelerator_connector.py [hunk 7]
```diff
@@ -918,10 +925,10 @@ def _mock_tpu_available(value):
         mock_tpu_available(monkeypatch, False)
         mock_ipu_available(monkeypatch, False)
         mock_hpu_available(monkeypatch, True)
-        monkeypatch.setattr(lightning.pytorch.accelerators.hpu.HPUAccelerator, "auto_device_count", lambda *_: 1)
+        monkeypatch.setattr(lightning_habana.HPUAccelerator, "auto_device_count", lambda *_: 1)
         connector = _AcceleratorConnector()
     assert isinstance(connector.accelerator, HPUAccelerator)
-    assert isinstance(connector.strategy, SingleHPUStrategy)
+    assert isinstance(connector.strategy, HPUSingleStrategy)
```
[the remark this answers is not in the document]
Comment (Borda) [reply tests/tests_pytorch/trainer/connectors/test_accelerator_connector.py:931]: You are right we're inconsistent in names between parallel and single strategies :(
so the remaining to fix is TPU as I do not see any other single strategy...

[the remark this answers is not in the document]
Comment (carmocca) [reply tests/tests_pytorch/trainer/connectors/test_accelerator_connector.py:931]: I mean that this change should be reverted. See it across the codebase

```python
$ grep -iIrn "class Single.*Strategy" src
src/lightning/fabric/strategies/single_tpu.py:23:class SingleTPUStrategy(SingleDeviceStrategy):
src/lightning/fabric/strategies/single_device.py:29:class SingleDeviceStrategy(Strategy):
src/lightning/pytorch/strategies/single_hpu.py:34:class SingleHPUStrategy(SingleDeviceStrategy):
src/lightning/pytorch/strategies/single_tpu.py:26:class SingleTPUStrategy(SingleDeviceStrategy):
src/lightning/pytorch/strategies/single_device.py:28:class SingleDeviceStrategy(Strategy):
```

**Comment by awaelchli** [comment]
Why is this error now suddenly showing up: 

> Should not import standalone package, all imports should be redirected to the unified package 

😠

**Comment by Borda** [comment]
> Why is this error now suddenly showing up:
> 
> > Should not import standalone package, all imports should be redirected to the unified package

good question and it shall be already appearing with #16921

**Comment by awaelchli** [comment]
The strange error only seems to occur in this group of tests 🤔 


```
Test PyTorch / pl-cpu (macOS-11, pytorch, 3.8, 1.13) 
Test PyTorch / pl-cpu (ubuntu-20.04, pytorch, 3.8, 1.13)
Test PyTorch / pl-cpu (windows-2022, pytorch, 3.8, 1.13) 
```
Which is anything outside the lightning meta package. I think it might be because the `lightning-habana` repo does not support importing things from `pytorch_lightning`, only from `lightning`. Fixing this is beyond my skills.

**Comment by Borda** [comment]
> Which is anything outside the lightning meta package. I think it might be because the `lightning-habana` repo does not support importing things from `pytorch_lightning`, only from `lightning`. Fixing this is beyond my skills.

addressing it in afa0b92bd09669f7f659ccefd0699af9fdaa112f

File: .github/workflows/ci-tests-pytorch.yml [hunk 8]
```diff
@@ -139,12 +140,18 @@ jobs:
...
-    - name: Adjust tests -> PL
+    - name: Adjust tests / env. -> PL
       if: ${{ matrix.pkg-name != 'lightning' }}
       run: |
         python .actions/assistant.py copy_replace_imports --source_dir="./tests" \
           --source_import="lightning.fabric,lightning.pytorch" \
           --target_import="lightning_fabric,pytorch_lightning"
+        # Lightning is dependency of Habana or other accelerators/integrations so in case we test PL we need to remove it
+        pip uninstall -y lightning
+        pip list
```
Comment (Borda) [inline .github/workflows/ci-tests-pytorch.yml:151]:
```suggestion
```

File: docs/source-pytorch/accelerators/hpu.rst [hunk 9]
```diff
@@ -1,40 +1,107 @@
-.. _hpu:
+:orphan:
```
Comment (Borda) [inline docs/source-pytorch/accelerators/hpu.rst:1]: @lantiga this is the dense one-ager in our docs

File: docs/source-pytorch/accelerators/hpu.rst [hunk 10]
```diff
@@ -1,40 +1,107 @@
...
+The TPC core is a VLIW SIMD processor with an instruction set and hardware tailored to serve training workloads efficiently.
+The Gaudi memory architecture includes on-die SRAM and local memories in each TPC and,
+Gaudi is the first DL training processor that has integrated RDMA over Converged Ethernet (RoCE v2) engines on-chip.
+
+On the software side, the PyTorch Habana bridge interfaces between the framework and SynapseAI software stack to enable the execution of deep learning models on the Habana Gaudi device.
+
+Gaudi offers a substantial price/performance advantage -- so you get to do more deep learning training while spending less.
+
+For more information, check out `Gaudi Architecture <https://docs.habana.ai/en/latest/Gaudi_Overview/Gaudi_Overview.html#gaudi-architecture>`__ and `Gaudi Developer Docs <https://developer.habana.ai>`__.
+
```
Comment (Borda) [inline docs/source-pytorch/accelerators/hpu.rst:25]:
```suggestion

----

Install integration
-------------------

.. code-block:: bash

    pip install lightning-habana -U

```

File: docs/source-pytorch/accelerators/hpu.rst [hunk 11]
```diff
@@ -1,40 +1,118 @@
-.. _hpu:
+:orphan:
 
 Accelerator: HPU training
 =========================
+**Audience:** Users looking to save money and run large models faster using single or multiple Gaudi devices.
```
Comment (Borda) [inline docs/source-pytorch/accelerators/hpu.rst:5]:
```suggestion

**Audience:** Users looking to save money and run large models faster using single or multiple Gaudi devices.
```

File: tests/tests_pytorch/trainer/connectors/test_accelerator_connector.py [hunk 12]
```diff
@@ -566,16 +568,17 @@ def test_unsupported_ipu_choice(monkeypatch):
 
 
 def mock_hpu_available(monkeypatch, value=True):
-    monkeypatch.setattr(lightning.pytorch.accelerators.hpu, "_HPU_AVAILABLE", value)
-    monkeypatch.setattr(lightning.pytorch.accelerators.hpu.HPUAccelerator, "is_available", lambda: value)
-    monkeypatch.setattr(lightning.pytorch.strategies.hpu_parallel, "_HPU_AVAILABLE", value)
-    monkeypatch.setattr(lightning.pytorch.strategies.single_hpu, "_HPU_AVAILABLE", value)
-    monkeypatch.setattr(lightning.pytorch.plugins.precision.hpu, "_HPU_AVAILABLE", value)
+    import lightning_habana
```
Comment (carmocca) [inline tests/tests_pytorch/trainer/connectors/test_accelerator_connector.py:571]: This should return if it's not available

Comment (Borda) [reply tests/tests_pytorch/trainer/connectors/test_accelerator_connector.py:571]:
```suggestion
    try:
        import lightning_habana
    except ModuleNotFoundError:
        return
```

**Comment by Borda** [comment]
some of the failures above are addresses in #17170

File: tests/tests_pytorch/accelerators/test_registry.py [hunk 13]
```diff
@@ -15,4 +15,8 @@
 
 
 def test_available_accelerators_in_registry():
-    assert AcceleratorRegistry.available_accelerators() == ["cpu", "cuda", "hpu", "ipu", "mps", "tpu"]
+    """Tests the accelerators available by default, not including external, third-party accelerators."""
+    available = set(AcceleratorRegistry.available_accelerators())
+    expected = {"cpu", "cuda", "ipu", "mps", "tpu"}
+    # Note: the registry is global, other tests may register new strategies as a side effect
+    assert expected.issubset(available)
```
Comment (akihironitta) [inline tests/tests_pytorch/accelerators/test_registry.py:22]: 👍

**Review by akihironitta — APPROVED** [review]
LGTM! Do we need a changelog entry for this change?

File: docs/source-pytorch/extensions/strategy.rst [hunk 14]
```diff
@@ -81,12 +81,6 @@ The below table lists all relevant strategies available in Lightning with their
    * - deepspeed
      - :class:`~lightning.pytorch.strategies.DeepSpeedStrategy`
      - Provides capabilities to run training using the DeepSpeed library, with training optimizations for large billion parameter models. :ref:`Learn more. <advanced/model_parallel:deepspeed>`
-   * - hpu_parallel
```
Comment (lantiga) [inline docs/source-pytorch/extensions/strategy.rst:84]: I would keep these, with a note that they live in a separate package

**Review by lantiga — CHANGES_REQUESTED** [review]
Just one comment and we're good to go

File: docs/source-pytorch/extensions/strategy.rst (hunk 14 above)
Comment (Borda) [reply docs/source-pytorch/extensions/strategy.rst:84]: resolved

**Review by lantiga — APPROVED** [review]
Looks great!

**Comment by carmocca** [comment]
Is this meant to be in the 2.0.1 release?
